src/token/api/src/types/account_identifier.rs

```diff
@@ -0,0 +1,295 @@
+use std::fmt::{Display, Formatter};
+use std::str::FromStr;
+
+use candid::{CandidType, Principal};
+use serde::de::Error;
+use serde::{de, Deserialize, Serialize};
+use sha2::{Digest, Sha224};
+
+pub static SUB_ACCOUNT_ZERO: Subaccount = Subaccount([0; 32]);
+static ACCOUNT_DOMAIN_SEPERATOR: &[u8] = b"\x0Aaccount-id";
+
+#[derive(Debug, Clone, Hash, PartialEq, Eq, PartialOrd, Ord, Copy)]
+pub struct AccountIdentifier {
+    pub hash: [u8; 28],
+}
+
+impl AccountIdentifier {
+    pub fn new(account: Principal, sub_account: Option<Subaccount>) -> Self {
+        let mut hash = Sha224::new();
+        hash.update(ACCOUNT_DOMAIN_SEPERATOR);
+        hash.update(account.as_slice());
+
+        let sub_account = sub_account.unwrap_or(SUB_ACCOUNT_ZERO);
+        hash.update(&sub_account.0[..]);
+        Self {
+            hash: hash.finalize().into(),
+        }
+    }
+
+    /// Generates anonymous account identifier.
+    pub fn anonymous() -> AccountIdentifier {
+        Self::new(Principal::anonymous(), None)
+    }
+
+    pub fn empty() -> AccountIdentifier {
+        AccountIdentifier { hash: [0; 28] }
+    }
+
+    pub fn from_hex(hex_str: &str) -> Result<AccountIdentifier, String> {
+        let hex: Vec<u8> = hex::decode(hex_str).map_err(|e| e.to_string())?;
+        Self::from_slice(&hex[..]).map_err(|err| match err {
+            // Since the input was provided in hex, return an error that is hex-friendly.
+            AccountIdParseError::InvalidLength(_) => format!(
+                "{} has a length of {} but we expected a length of 64 or 56",
```

Comment:
either or in between? In any case those could be instances of `TryFrom<&str>` (I wish blanked impl from core didn't prevent writing covariant `TryFrom<impl AsRef<str>>`)

Reply:
I don't quite understand by what you mean either or in ?

Reply:
"{} has a length of {} but we expected a length of either 64 or 56",